Machine Learning/Class Assigment Objectives/classification-class-assignment-objective.ipynb

```diff
@@ -0,0 +1,631 @@
+    "# data wrangling and visualization\n",
+    "from matplotlib import pylab as plt\n",
+    "import pandas as pd\n",
+    "from sklearn.datasets import make_classification\n",
+    "from sklearn import metrics\n",
+    "import seaborn as sns\n",
+    "from IPython.display import Markdown, display\n",
+    "\n",
+    "# other utils\n",
+    "import json\n",
+    "import pprint\n",
+    "import time"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "hydraulic-google",
+   "metadata": {},
+   "source": [
+    "# Choosing the class assignment objective\n",
+    "\n",
+    "---\n",
+    "#### TL;DR\n",
+    "\n",
+    "* The default objective `maximize_minimum_recall` should work well for most cases.\n",
+    "* You should explicitly use the `maximize_minimum_recall` objective if you have an unbalanced dataset and you are interested in high recall on the rare class.\n",
+    "* You should use the `maximize_accuracy` objective if you have a balanced dataset and you are interested in high accuracy and precision.\n",
+    "---\n",
+    "\n",
+    "When creating a classification analysis job in Elasticsearch, you can specify the [`class_assignment_objective`](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/put-dfanalytics.html#ml-put-dfanalytics-request-body) using one of two parameters: `maximize_accuracy` and `maximize_minimum_recall`. Depending on your goal and your dataset, you may prefer a different option. \n",
+    "\n",
+    "When evaluating results of a classification job one can be interested in one of three different metrics: [precision, recall, and accuracy](https://en.wikipedia.org/wiki/Precision_and_recall). Let's recap what they do, using an example of a test for a not further specified virus.  \n",
+    "- **Precision** is the fraction of relevant instances among the retrieved instances. Precision gives us two numbers: \n",
+    "    1. What is the percentage of healthy people among those that were identified as \"healthy\" by the test and \n",
+    "    2. What is the percentage of infected people among those that were identified as \"infected\" by the test.\n",
+    "- **Recall** is the fraction of relevant instances that were retrieved. Again, recall gives us two numbers: \n",
+    "    1. What is the percentage of people identified as \"healthy\" by the test is among all healthy people and \n",
+    "    2. What is the percentage of people identified as \"infected\" is among all infected people.\n",
+    "- **Accuracy** is the fraction of relevant instances among all instances. It gives us a single number: the percentage of correctly identified healthy and infected people among all tested people.\n",
+    "\n",
+    "There is a trade-off between maximizing precision and recall. For example, labeling all people as \"infected\" will get us 100% but recall, since we identified all infected people in the sample, but only a very small precision since we misclassified many healthy people as \"infected\" as well.\n",
```

Comment:
I think this should maybe be tweaked slightly. It assumes that there is one important class for which you care about recall, but this wouldn't always be the case...
```suggestion
    "There is a trade-off between maximizing precision and recall. For example, labeling all people as \"infected\" will get us 100% but recall for infected people, since we identified all infected people in the sample, but only a very small precision since we misclassified many healthy people as \"infected\" as well.\n",
```